.github/workflows/integration-tests.yaml

```diff
@@ -0,0 +1,107 @@
+---
+name: Integration Tests
+on:
+  pull_request:
+    paths:
+      - "action.yaml"
+      - ".github/workflows/integration-tests.yaml"
+  push:
+    branches:
+      - main
+    paths:
+      - "action.yaml"
+      - ".github/workflows/integration-tests.yaml"
+
+jobs:
+  test:
+    name: Test ${{ matrix.test.title }}
+    # These permissions are needed to:
+    # - Checkout the repo
+    permissions:
+      contents: read
+      packages: write
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        test:
+          - title: Merge Commit
+            commit-sha: ${{ github.sha }}
+            from-scratch: true
+          - title: Head Commit
+            commit-sha: ${{ github.event.pull_request.head.sha }}
+    steps:
+      - name: Job started at
+        id: job-started
+        run: |
+          job_started_at="$(date --utc --iso-8601=seconds)"
+          echo "at=$job_started_at" | tee -a "$GITHUB_OUTPUT"
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ matrix.test.commit-sha }}
+      - name: Log in to the Container registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ github.token }}
+      - uses: ./
+        id: build
+        with:
+          image-repository: ghcr.io/beacon-biosignals/temporary/whalesay
+          context: test
+          build-args: |
+            DEBIAN_VERSION=12.9
+          from-scratch: ${{ matrix.test.from-scratch || 'false' }}
+      - name: Validate image works
+        run: |
+          docker pull "${{ steps.build.outputs.image }}"
+          output="$(docker run "${{ steps.build.outputs.image }}")"
+          if [[ "$(wc -l <<<"$output")" -lt 14 ]]; then
+              echo "$output"
+              exit 1
+          fi
+          debian_version="$(docker run --entrypoint=/bin/cat "${{ steps.build.outputs.image }}" /etc/debian_version)"
+          [[ "$debian_version" == "12.9" ]] || exit 2
+      - name: Layer created at
+        id: layer-created
+        run: |
+          layer_created_at="$(docker run --entrypoint=/bin/cat "${{ steps.build.outputs.image }}" /etc/layer-created-at)"
+          echo "at=$layer_created_at" | tee -a "$GITHUB_OUTPUT"
+      - name: Validate layer caching
+        if: ${{ matrix.test.from-scratch == false }}
+        run: |
+          [[ "$(date -d "$layer_created_at" +%s)" -lt "$(date -d "$job_started_at" +%s)" ]] || exit 1
+        env:
+          job_started_at: ${{ steps.job-started.outputs.at }}
+          layer_created_at: ${{ steps.layer-created.outputs.at }}
```

Comment:
Will fail on main the first time it is run. I'll just re-run the workflow for that scenario